src/main/java/org/assertj/core/internal/Maps.java

```diff
@@ -406,13 +451,48 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
     if (actual.isEmpty() && entries.length == 0) return;
     failIfEntriesIsEmptyEmptySinceActualIsNotEmpty(info, actual, entries);
 
-    Set<Entry<? extends K, ? extends V>> notFound = new LinkedHashSet<>();
-    Set<Entry<? extends K, ? extends V>> notExpected = new LinkedHashSet<>();
-    compareActualMapAndExpectedEntries(actual, entries, notExpected, notFound);
-    if (!notFound.isEmpty() || !notExpected.isEmpty())
+    Set<Entry<? extends K, ? extends V>> notFound = getNotFoundEntries(actual, entries);
+    Set<Entry<K, V>> notExpected = getNotExpectedEntries(actual, entries);
+
+    if (!(notFound.isEmpty() && notExpected.isEmpty()))
       throw failures.failure(info, shouldContainOnly(actual, entries, notFound, notExpected));
   }
 
+  private static <K, V> Set<Entry<? extends K, ? extends V>> getNotFoundEntries(Map<K, V> actual,
+                                                                                Entry<? extends K, ? extends V>[] entries) {
+    Set<Entry<? extends K, ? extends V>> notFound = new LinkedHashSet<>();
+    for (Entry<? extends K, ? extends V> entry : entries) {
+      if (!containsEntry(actual, entry)) notFound.add(entry);
+    }
+    return notFound;
+  }
+
+  private static <K, V> Set<Entry<K, V>> getNotExpectedEntries(Map<K, V> actual, Entry<? extends K, ? extends V>[] entries) {
+    Set<Entry<K, V>> notExpected = new LinkedHashSet<>();
+    for (Entry<K, V> entry : mapWithoutExpectedEntries(actual, entries).entrySet()) {
+      MapEntry<K, V> mapEntry = entry(entry.getKey(), entry.getValue());
+      notExpected.add(mapEntry);
+    }
+    return notExpected;
+  }
+
+  private static <K, V> Map<K, V> mapWithoutExpectedEntries(Map<K, V> actual, Entry<? extends K, ? extends V>[] expectedEntries) {
+    try {
+      Map<K, V> clonedMap = clone(actual);
+      for (Entry<? extends K, ? extends V> expectedEntry : expectedEntries) {
+        clonedMap.remove(expectedEntry.getKey(), expectedEntry.getValue());
+      }
+      return clonedMap;
+    } catch (NoSuchMethodException | UnsupportedOperationException e) {
```

Comment:
should we also catch the `IllegalStateException` thrown by `clone(Map<K, V> map)`?

Reply:
I was wondering about the same.

Catching `IllegalStateException` (or even propagate the checked exceptions and catch them here) would make the assertion more robust.

On the other side, we may use `LinkedHashMap` for implementations that could be handled in a better way. I can imagine these would be exotic map implementations. It could be worth getting the knowledge about them and add proper tests.

Currently, I would take the risk, i.e., no catch for `IllegalStateException`.

Reply:
fine by me